File: MiscChallenges/Challenges/Programming Challenges/Collatz.cs
using System;
using System.Linq;
using static System.Console;
using static System.Math;

namespace MiscChallenges.Challenges
{
	public static partial class ChallengeClass
	{
		[Challenge("ProgChallenges", "The 3n + 1 Problem",
			"http://www.programming-challenges.com/pg.php?page=downloadproblem&probid=110101&format=html")]
		public class Collatz : IChallenge
		{
			public void Solve()
			{
				while (true)
				{
					var nextCase = GetCase();
					if (nextCase == null)
					{
						break;
					}

					var low = Min(nextCase.Item1, nextCase.Item2);
					var high = Max(nextCase.Item1, nextCase.Item2);
					var c = Enumerable.Range(low, high - low + 1).
						Select(CollatzCount).
						Max();
                    WriteLine(@"{0} {1} {2}", nextCase.Item1, nextCase.Item2, c);
				}
			}

			private int CollatzCount(int i)
			{
				long v = i;
				var count = 1;

				while (v != 1)
				{
					count++;
					if ((v & 1) == 0)
					{
						v = v >> 1;
					}
					else
					{
						v = (v << 1) + v + 1;
					}
				}
				return count;
			}

			private Tuple<int, int> GetCase()
			{
				var line = ReadLine();
				if (line == null)
				{
					return null;
				}
				var vals = line.Split(' ').
					Select(int.Parse).
					ToList();
				return new Tuple<int, int>(vals[0], vals[1]);
			}

			public string RetrieveSampleInput()
			{
				return @"
1 10
100 200
201 210
900 1000
1 1
10 1
210 201
113383 113383
999999 1
";
			}

			public string RetrieveSampleOutput()
			{
				return @"
1 10 20
100 200 125
201 210 89
900 1000 174
1 1 1
10 1 20
210 201 89
113383 113383 248
999999 1 525
";
			}
		}
	}
}

File: CPP Challenges/Collatz.cpp
#ifdef CHALLENGE_RUNNER
#include "stdafx.h"
#include "Challenge.h"
#else
#define prolog(a,b,c,d)
#define epilog
#define sampleInput char *_dummy1_ =
#define sampleOutput char *_dummy2_ =
#endif

#include<iostream>
#include<math.h>

prolog("ProgChallenges", Collatz, "The 3n+1 Problem - cpp", "http://www.programming-challenges.com/pg.php?page=downloadproblem&probid=110101&format=html")

int CollatzCount(int a)
{
	long long v = a;
	auto count = 1;

	while (v != 1)
	{
		count++;
		if ((v & 1) == 0)
		{
			v = v / 2;
		}
		else
		{
			v = 3 * v + 1;
		}
	}
	return count;
}

using namespace std;
int main()
{
	long n1, n2;
	int maxSteps;

	while (true)
	{
		cin >> n1 >> n2;
		if (cin.eof())
		{
			break;
		}
		maxSteps = -1;
		int low = min(n1, n2);
		auto high = max(n1, n2);
		for (auto i = low; i <= high; i++)
		{
			auto cycleLength = CollatzCount(i);
			maxSteps = max(cycleLength, maxSteps);
		}
		cout << n1 << ' ' << n2 << ' ' << maxSteps << endl;
	}

	return 0;
}

sampleInput
R"(
1 10
100 200
201 210
900 1000
1 1
10 1
210 201
113383 113383
999999 1
)";

sampleOutput
R"(
1 10 20
100 200 125
201 210 89
900 1000 174
1 1 1
10 1 20
210 201 89
113383 113383 248
999999 1 525
)";

epilog
